server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -593,6 +593,31 @@ public boolean isSystemContext() {
         return threadLocal.get().isSystemContext;
     }
 
+    /**
+     * Remove unwanted and unneeded headers from the thread context. Does not store prior context.
+     */
+    public void sanitizeHeaders() {
+        final ThreadContextStruct originalContext = threadLocal.get();
+        final Map<String, String> newRequestHeaders = new HashMap<>(originalContext.requestHeaders);
+
+        newRequestHeaders.entrySet()
+            .removeIf(
+                entry -> entry.getKey().equalsIgnoreCase("authorization")
+                    || entry.getKey().equalsIgnoreCase("es-secondary-authorization")
+                    || entry.getKey().equalsIgnoreCase("ES-Client-Authentication")
+            );
+
+        final ThreadContextStruct newContext = new ThreadContextStruct(
+            newRequestHeaders,
+            originalContext.responseHeaders,
+            originalContext.transientHeaders,
+            originalContext.isSystemContext,
+            originalContext.warningHeadersSize
+        );
+        threadLocal.set(newContext);
```

Comment:
It's only rarely that we set the thread-local thread-context outside of a try-with-resource block.
Setting it like that is relatively perilous as it affects everything else after that, up the call stack too, until a context stash or the thread finishing. Hence it's difficult to asses the code coverage of the thread-local change.

I think we can achieve the same sanitize outcome with a try-with-resource block ala https://github.com/elastic/elasticsearch/blob/c3a1cb34294e902a9f46d7e840ea09965019f456/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L232 .
This way the code in `AuthenticationService#authenticate` doesn't have to consider that calling the argument listener can potentially affect thread-context headers.

What do you think?

Reply:
We chatted abit in person and will leave the implementation as-is. This approach could work, but has an undesirable effect that when the try-with-resources get closed it will place back the old un-sanitized headers. Meaning that the headers are only sanitized for a given a scope which is conditional on the execution stack. We want them completed removed after we know they should no longer be needed. That may complicate future designs if the definition of no longer needed changes, but we can address if/when that happens. (global variables are fun :/)

Reply:
Another alternative is to filter out these headers when sending request across nodes, i.e. [SecurityServerTransportInterceptor#interceptSender](https://github.com/elastic/elasticsearch/blob/ef7aaec815baa7111bf21379f2d316b800ccc38b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java#L77), where we are already doing a bunch of header transformation. It does mean the headers are kept alive on the local node. But since we mainly want to prevent these headers from reaching another node, this might be good enough? It also avoids having to mutate the ThreadContext in a less conventional way.

I do *not* think this approach is necessarily better. Clearing the headers immediately after their usage is likely conceptually better. Just in case it has not been discussed, it might worth some consideration. Thanks!

Reply:
That is a possibility, but I would prefer to sanitize these as soon as possible.